src/EGraphs/saturation.jl

```diff
@@ -87,14 +87,14 @@ function eqsat_search!(
 
       for i in ids_left
         cansearch(scheduler, rule_idx, i) || continue
-        n_matches += rule.ematcher_left!(g, rule_idx, i, rule.stack, ematch_buffer)
+        n_matches += rule.ematcher_left!(g, rule_idx, i, get_local_stack(), ematch_buffer)
```

Comment:
Are we sure the compiler is doing constant propagation for `get_local_stack()`?

I would move this to a `stack = get_local_stack()` above  `for (rule_idx, rule) in enumerate(theory)` to avoid computing the stack retrieval for each rule and for each id in the e-graph, could be very costly, and should still protect from the race condition